pkg/controller/certificatesigningrequests/controller.go

```diff
@@ -0,0 +1,169 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificatesigningrequests
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	certificatesclient "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	certificateslisters "k8s.io/client-go/listers/certificates/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmlisters "github.com/jetstack/cert-manager/pkg/client/listers/certmanager/v1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+)
+
+const (
+	ControllerName = "certificatesigningrequests"
+)
+
+var keyFunc = controllerpkg.KeyFunc
+
+type Signer interface {
+	Sign(context.Context, *certificatesv1.CertificateSigningRequest, cmapi.GenericIssuer) error
+}
+
+type Controller struct {
+	helper issuer.Helper
+
+	//// clientset used to update cert-manager API resources
+	//cmClient cmclient.Interface
+	certClient certificatesclient.CertificateSigningRequestInterface
+
+	csrLister certificateslisters.CertificateSigningRequestLister
+
+	queue workqueue.RateLimitingInterface
+
+	// logger to be used by this controller
+	log logr.Logger
+
+	// used to record Events about resources to the API
+	recorder record.EventRecorder
+
+	// the signer kind to react to when a certificate signing request is synced
+	signerType string
+
+	issuerLister        cmlisters.IssuerLister
+	clusterIssuerLister cmlisters.ClusterIssuerLister
+
+	// Signer to call sign function
+	signer Signer
+
+	//// used for testing
+	clock clock.Clock
+
+	//reporter *util.Reporter
+}
+
+func New(signerType string, signer Signer) *Controller {
+	return &Controller{
+		signerType: signerType,
+		signer:     signer,
+	}
+}
+
+func (c *Controller) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// construct a new named logger to be reused throughout the controller
+	c.log = logf.FromContext(ctx.RootContext, ControllerName)
+
+	// create a queue used to queue up items to be processed
+	c.queue = workqueue.NewNamedRateLimitingQueue(controllerpkg.DefaultItemBasedRateLimiter(), ControllerName)
+
+	issuerInformer := ctx.SharedInformerFactory.Certmanager().V1().Issuers()
+	c.issuerLister = issuerInformer.Lister()
```

Comment:
How is namespacing being handled here given CSRs are non-namespace scoped?

Reply:
Though not in this PR, the intention is to use the SubjectAccessReview gate described in the design doc [here](https://github.com/jetstack/cert-manager/pull/3646/files#diff-456aff4025c7d34e61065ec6ac3bbc80ca147b7d7b46f05c9caffa0fd5633641R141). (Pending agreement on which verbs etc to use.).